### PR TITLE
Various changes after splitting consensus API

### PR DESCRIPTION
**Pull request by coot**
These changes I came up with when I started updating `cardano-node` repository.

Related to #1950

- supported versions should be a non empty list
- ClientCodecs: node-to-client protocol codecs
- defaultCodec: don't impose IOLike where MonadST is enough
- added localStateQueryClientNull
- Added utililty functions for folding multiple Versions
- NodeToClient.versionedNodeToClientProtocols

### CODE REVIEW

**Comment by coot** [comment]
The clients depend on `TopLeevelConfig` because of `nodeNetworkMagic`.

**Comment by coot** [comment]
bors merge

File: ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToClient.hs [hunk 1]
```diff
@@ -161,6 +173,42 @@ defaultCodecs _cfg _version = Codecs {
...
     }
 
+
+-- | Protocol codecs for the node-to-client protocols which serialise
+-- / deserialise blocks in /chain-sync/ protocol.
+--
+clientCodecs :: forall m blk. (RunNode blk, MonadST m)
+              => BlockConfig         blk
+              -> NodeToClientVersion blk
+              -> ClientCodecs blk m
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToClient.hs:183]:
```suggestion
clientCodecs :: forall m blk. (RunNode blk, MonadST m)
             => BlockConfig         blk
             -> NodeToClientVersion blk
             -> ClientCodecs blk m
```

**Comment by coot** [comment]
bors merge

**Pull request merged**